services/payments/statemachine_zebpay.go

```diff
@@ -62,7 +65,16 @@ func (zm *ZebpayMachine) Pay(ctx context.Context) (*paymentLib.AuthenticatedPaym
 		case zebpay.TransferPendingCode:
 			// Set backoff without changing status
 			zm.backoffFactor = zm.backoffFactor * 2
+		case zebpay.TransferNotFoundCode:
+			// The transaction was never received by zebpay. We should not have gotten into this
+			// condition unless zebpay accepted this transaction, but failed to persist the record.
+			// We'll need to retry the submission and come back around to check status, so do not
+			// SetNextState except those that get set as part of the submit function
+			return zm.transaction, zm.submit(ctx, entry, to)
```

Comment:
This is a bit sinister – the `submit` call is hidden. It seems to make sense to optimise for readability and perform it on its own, and return based on the result.